src/cellrank/kernels/_base_kernel.py

```diff
@@ -463,6 +464,88 @@ def _reuse_cache(self, expected_params: Dict[str, Any], *, time: Optional[Any] =
             self._params = expected_params
         # fmt: on
 
+    def _get_mask(self, series, subset):
+        if isinstance(subset, str):
+            subset = [subset]
+        return series.isin(subset)
+
+    def get_boundary(self, source: str, target: str, cluster_key: str, graph: str = "distances"):
+        """Identify source observations at boundary to target cluster."""
+        source_obs_mask = self._get_mask(series=self.adata.obs[cluster_key], subset=source)
+        target_obs_mask = self._get_mask(series=self.adata.obs[cluster_key], subset=target)
+        obs_ids = np.arange(0, self.adata.n_obs)
+
+        source_ids = obs_ids[source_obs_mask]
+        adj_mat_subset = self.adata.obsp[graph][source_ids, :]
+
+        boundary_ids = []
+        for row_id, row in enumerate(adj_mat_subset):
+            _obs_mask = row.A.squeeze().astype(bool)
+            _obs_mask = _obs_mask & target_obs_mask
+
+            if _obs_mask.any():
+                boundary_ids.append(source_ids[row_id])
+
+        return boundary_ids
+
+    def get_empirical_velocity_field(self, boundary_ids, target_obs_mask, rep: str, graph: str = "distances"):
```

Comment:
@michalk8, `target_obs_mask` is a boolean mask over observations to select observations of the target cluster. Do you see a nicer implementation than simply passing it to the function? An alternative could be to pass the name of the target cluster and compute the mask in the function itself.

Reply:
Done.